pkg/ride/generate/internal/constructors_generation.go

```diff
@@ -0,0 +1,283 @@
+		verInfo := verInfos[ver]
+		if verInfo == nil {
+			verInfo = newVersionInfo(ver)
+			verInfos[ver] = verInfo
+		}
+
+		for _, name := range verInfo.removedStructs {
+			delete(existingStructs, name)
+		}
+		for _, structInfo := range verInfo.newStructs {
+			existingStructs[structInfo.rideName] = structInfo
+		}
+
+		verInfo.newStructs = make([]constructorStructInfo, 0, len(existingStructs))
+		for _, structInfo := range existingStructs {
+			verInfo.newStructs = append(verInfo.newStructs, structInfo)
+		}
+	}
+
+	return nil
+}
+
+func constructorsHandleRideObject(cd *Coder, obj rideObject) error {
+	if obj.SkipConstructor {
+		return nil
+	}
+
+	for _, act := range obj.Actions {
+		constructorName := constructorName(act)
+		cd.Line("func %s(_ environment, args_ ...rideType) (rideType, error) {", constructorName)
+
+		arguments, err := extractConstructorArguments(act.Fields)
+		if err != nil {
+			panic(errors.Wrap(err, obj.Name).Error())
+		}
+
+		cd.Line("if err := checkArgs(args_, %d); err != nil {", len(arguments))
+		cd.Line("return nil, errors.Wrap(err, \"%s\")", constructorName)
+		cd.Line("}")
+		cd.Line("")
+
+		for i, arg := range arguments {
+			varName := argVarName(arg)
+
+			if len(arg.Types) == 1 {
+				info := arg.Types[0]
+				cd.Line("%s, ok := args_[%d].(%s)", varName, i, info)
+				cd.Line("if !ok {")
+				cd.Line("return nil, errors.Errorf(\"%s: unexpected type '%%s' for %s\", args_[%d].instanceOf())", constructorName, varName, i)
+				cd.Line("}")
+
+				if listInfo, ok := info.(*listTypeInfo); ok {
+					cd.Line("// checks for list elements")
+					checkListElementsTypes(cd, constructorName, varName, listInfo)
+				}
+			} else {
+				cd.Line("var %s rideType", varName)
+				cd.Line("switch v := args_[%d].(type) {", i)
+				onelineTypes := make([]string, 0, len(arg.Types))
+				for _, t := range arg.Types {
+					onelineTypes = append(onelineTypes, t.String())
+				}
+				cd.Line("case %s:", strings.Join(onelineTypes, ", "))
+				cd.Line("%s = v", varName)
+				cd.Line("default:")
+				cd.Line("return nil, errors.Errorf(\"%s: unexpected type '%%s' for %s\", args_[%d].instanceOf())", constructorName, varName, i)
+				cd.Line("}")
+			}
+			cd.Line("")
+		}
+
+		argsStr := make([]string, len(act.Fields))
+		for i, arg := range act.Fields {
+			varName := argVarName(arg)
+			if arg.ConstructorOrder == -1 {
+				cd.Line("// default value for %s", varName)
+				cd.Line("var %s %s", varName, getType(arg.Types))
+			}
+			argsStr[i] = varName
+		}
+
+		cd.Line("")
+		cd.Line("return %s(%s), nil", rideActionConstructorName(act), strings.Join(argsStr, ", "))
+		cd.Line("}")
+		cd.Line("")
+
+		if act.Deleted != nil {
+			verInfos.addRemoved(*act.Deleted, obj.Name)
+		}
+
+		verInfos.addNewStruct(act.LibVersion, constructorStructInfo{
+			rideName:   obj.Name,
+			goName:     constructorName,
+			argsNumber: len(arguments),
+		})
+	}
+
+	return nil
+}
+
+var verInfos = versionInfos{}
```

Comment:
For which purpose you've decided to use global variable? Is it necessary to be global? On my mind it looks a bit weird and (of course you know) as a rule of thumb global variables should be avoided.

Reply:
We need to parse json config and build `versionInfos` for further using it in generating functions catalogues, but it's evaluating in the same binary, so we need to store it somewhere for functions catalogue.

But agree that using global variables is a bad pattern, so moved the whole implementation to `vinfo` package and changed using this global variable to a Singletone pattern.